File: 2024/day17/go/solution.go
// Day 17: Chronospatial Computer - 3-bit VM emulator
package main

import (
	"fmt"
	"os"
	"regexp"
	"slices"
	"strconv"
	"strings"
)

func parseInput(text string) (int64, int64, int64, []int) {
	lines := strings.Split(strings.TrimSpace(text), "\n")

	reA := regexp.MustCompile(`Register A: (\d+)`)
	reB := regexp.MustCompile(`Register B: (\d+)`)
	reC := regexp.MustCompile(`Register C: (\d+)`)
	reProg := regexp.MustCompile(`Program: ([\d,]+)`)

	// Errors ignored - input is trusted AoC format
	a, _ := strconv.ParseInt(reA.FindStringSubmatch(lines[0])[1], 10, 64)
	b, _ := strconv.ParseInt(reB.FindStringSubmatch(lines[1])[1], 10, 64)
	c, _ := strconv.ParseInt(reC.FindStringSubmatch(lines[2])[1], 10, 64)

	progStr := reProg.FindStringSubmatch(lines[4])[1]
	progParts := strings.Split(progStr, ",")
	program := make([]int, len(progParts))
	for i, p := range progParts {
		val, _ := strconv.Atoi(p) // Error ignored - trusted input
		program[i] = val
	}

	return a, b, c, program
}

func runProgram(a, b, c int64, program []int) []int {
	ip := 0
	var output []int

	combo := func(operand int) int64 {
		switch operand {
		case 0, 1, 2, 3:
			return int64(operand)
		case 4:
			return a
		case 5:
			return b
		case 6:
			return c
		default:
			panic(fmt.Sprintf("Invalid combo operand: %d", operand))
		}
	}

	for ip < len(program) {
		opcode := program[ip]
		operand := program[ip+1]

		switch opcode {
		case 0: // adv - A = A >> combo
			a = a >> combo(operand)
		case 1: // bxl - B = B XOR literal
			b = b ^ int64(operand)
		case 2: // bst - B = combo % 8
			b = combo(operand) & 7
		case 3: // jnz - jump if A != 0
			if a != 0 {
				ip = operand
				continue
			}
		case 4: // bxc - B = B XOR C
			b = b ^ c
		case 5: // out - output combo % 8
			output = append(output, int(combo(operand)&7))
		case 6: // bdv - B = A >> combo
			b = a >> combo(operand)
		case 7: // cdv - C = A >> combo
			c = a >> combo(operand)
		}

		ip += 2
	}

	return output
}

func part1(text string) string {
	a, b, c, program := parseInput(text)
	output := runProgram(a, b, c, program)

	strs := make([]string, len(output))
	for i, v := range output {
		strs[i] = strconv.Itoa(v)
	}
	return strings.Join(strs, ",")
}

func part2(text string) int64 {
	_, b, c, program := parseInput(text)

	// The program loops, outputting one digit per iteration, dividing A by 8 each time.
	// We need to find A such that output == program.
	// Work backwards from the last digit - build A 3 bits at a time.

	var search func(targetIdx int, currentA int64) (int64, bool)
	search = func(targetIdx int, currentA int64) (int64, bool) {
		if targetIdx < 0 {
			return currentA, true
		}

		// Try all 8 possible 3-bit values for this position
		for bits := int64(0); bits < 8; bits++ {
			candidateA := (currentA << 3) | bits
			if candidateA == 0 && targetIdx == len(program)-1 {
				continue // A can't be 0 at start (would halt immediately without output)
			}

			output := runProgram(candidateA, b, c, program)

			// Check if output matches the suffix of the program
			expected := program[targetIdx:]
			if slices.Equal(output, expected) {
				result, found := search(targetIdx-1, candidateA)
				if found {
					return result, true
				}
			}
		}

		return 0, false
	}

	result, _ := search(len(program)-1, 0)
	return result
}

func main() {
	data, err := os.ReadFile("../input.txt")
	if err != nil {
		panic(err)
	}
	text := string(data)

	fmt.Println("Part 1:", part1(text))
	fmt.Println("Part 2:", part2(text))
}

File: 2024/day17/ruby/solution.rb
#!/usr/bin/env ruby
# Day 17: Chronospatial Computer - 3-bit VM emulator

def parse_input(text)
  lines = text.strip.split("\n")
  a = lines[0].match(/Register A: (\d+)/)[1].to_i
  b = lines[1].match(/Register B: (\d+)/)[1].to_i
  c = lines[2].match(/Register C: (\d+)/)[1].to_i
  program = lines[4].match(/Program: ([\d,]+)/)[1].split(',').map(&:to_i)
  [a, b, c, program]
end

def run_program(a, b, c, program)
  ip = 0
  output = []

  combo = ->(operand) {
    case operand
    when 0..3 then operand
    when 4 then a
    when 5 then b
    when 6 then c
    else raise "Invalid combo operand: #{operand}"
    end
  }

  while ip < program.size
    opcode = program[ip]
    operand = program[ip + 1]

    case opcode
    when 0  # adv - A = A >> combo
      a = a >> combo.call(operand)
    when 1  # bxl - B = B XOR literal
      b = b ^ operand
    when 2  # bst - B = combo % 8
      b = combo.call(operand) & 7
    when 3  # jnz - jump if A != 0
      if a != 0
        ip = operand
        next
      end
    when 4  # bxc - B = B XOR C
      b = b ^ c
    when 5  # out - output combo % 8
      output << (combo.call(operand) & 7)
    when 6  # bdv - B = A >> combo
      b = a >> combo.call(operand)
    when 7  # cdv - C = A >> combo
      c = a >> combo.call(operand)
    end

    ip += 2
  end

  output
end

def part1(text)
  a, b, c, program = parse_input(text)
  output = run_program(a, b, c, program)
  output.join(',')
end

def part2(text)
  _, b, c, program = parse_input(text)

  # The program loops, outputting one digit per iteration, dividing A by 8 each time.
  # We need to find A such that output == program.
  # Work backwards from the last digit - build A 3 bits at a time.

  search = ->(target_idx, current_a) {
    return current_a if target_idx < 0

    # Try all 8 possible 3-bit values for this position
    (0..7).each do |bits|
      candidate_a = (current_a << 3) | bits
      # A can't be 0 at start (would halt immediately without output)
      next if candidate_a == 0 && target_idx == program.size - 1

      output = run_program(candidate_a, b, c, program)

      # Check if output matches the suffix of the program
      expected = program[target_idx..]
      if output == expected
        result = search.call(target_idx - 1, candidate_a)
        return result if result
      end
    end

    nil
  }

  search.call(program.size - 1, 0)
end

if __FILE__ == $0
  input_path = File.join(File.dirname(__FILE__), '..', 'input.txt')
  text = File.read(input_path)

  puts "Part 1: #{part1(text)}"
  puts "Part 2: #{part2(text)}"
end
